demos/programs/drag_and_drop/simpledrop.man
.\" $XConsortium: simpledrop.man /main/4 1995/07/17 10:46:47 drk $
.\" Motif
.\"
.\" Copyright (c) 1987-2012, The Open Group. All rights reserved.
.\"
.\" These libraries and programs are free software; you can
.\" redistribute them and/or modify them under the terms of the GNU
.\" Lesser General Public License as published by the Free Software
.\" Foundation; either version 2.1 of the License, or (at your option)
.\" any later version.
.\"
.\" These libraries and programs are distributed in the hope that
.\" they will be useful, but WITHOUT ANY WARRANTY; without even the
.\" implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
.\" PURPOSE. See the GNU Lesser General Public License for more
.\" details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with these librararies and programs; if not, write
.\" to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
.\" Floor, Boston, MA 02110-1301 USA
...\" 
...\" 
...\" HISTORY
.TH simpledrop 1X MOTIF "Demonstration programs"
.SH NAME
\*Lsimpledrop\*O\(emis a simple demonstration of the drag and drop 
system.  Unlike DNDDemo, simpledrop is a minimal example.
.SH SYNOPSIS
.sS
\*Lsimpledrop\*O
.sE
.SH DESCRIPTION
\*Lsimpledrop\*O
shows off the most basic features in Motif Drag and Drop.  It also 
serves as a programming example.




.attic/ChangeLog
2017-11-10 11:49:10 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1665 (HEAD, origin/master, origin/HEAD, master)

2017-10-30 23:04:15 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1651

2017-10-30 22:22:31 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1650

2017-10-30 21:47:21 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1652

2017-08-31 06:28:27 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1655

2017-08-28 05:33:53 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1625

2017-08-26 06:12:14 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1656

2017-08-23 04:10:55 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1658

2017-08-22 06:05:00 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1659

2017-08-22 03:27:36 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1660

2017-03-28 01:40:33 +0300 Oleksiy Chernyavskyy 

	* updated version and release data

2017-03-28 01:32:30 +0300 Oleksiy Chernyavskyy 

	* removed redundant includes

2017-03-28 01:11:59 +0300 Oleksiy Chernyavskyy 

	* removed bugfix 1565. See details on http://bugs.motifzone.net/show_bug.cgi?id=1565

2016-11-09 07:19:53 +0200 Alex 

	* obsolete malloc.h changed on stdlib.h

2016-06-11 05:03:21 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-11 05:02:55 +0300 Oleksiy Chernyavskyy 

	* updated release 2.3.6 information

2016-06-11 04:57:54 +0300 Oleksiy Chernyavskyy 

	* fixid parallel building (make -j) error

2016-06-11 04:57:20 +0300 Oleksiy Chernyavskyy 

	* removed automatic configure start from autogen.sh

2016-06-11 04:26:48 +0300 Oleksiy Chernyavskyy 

	* minor fix

2016-06-11 04:21:53 +0300 Oleksiy Chernyavskyy 

	* removed extra dist Imakefile from all Makefile.am files

2016-06-11 04:06:37 +0300 Oleksiy Chernyavskyy 

	* Purified copyright headers (Motif is LGPL covered). Removed many Imakefile's and Imake-related stuff

2016-06-10 23:40:43 +0300 Oleksiy Chernyavskyy 

	* updated release notes and version 2.3.5 -> 2.3.6

2016-06-10 23:27:56 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-10 23:15:00 +0300 Oleksiy Chernyavskyy 

	* bug 1654 fix

2016-06-10 21:58:16 +0300 Oleksiy Chernyavskyy 

	* old patches. need review

2016-03-29 03:48:59 +0300 Oleksiy Chernyavskyy 

	* removed obsolete files

2016-03-17 04:23:17 +0200 Oleksiy Chernyavskyy 

	* release 2.3.5 (tag: release_2_3_5)

2016-03-16 23:07:12 +0200 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-03-16 23:06:24 +0200 Oleksiy Chernyavskyy 

	* added mkchangelog script for automated generation of the ChangeLog

2016-03-16 00:46:49 +0200 Oleksiy Chernyavskyy 

	* Reimplemented bugfix 1565

2016-02-26 03:36:51 +0200 Oleksiy Chernyavskyy 

	* Updated RPM creation instructions

2014-08-20 15:45:41 +0300 Oleksiy Chernyavskyy 

	* fixed bug 1636

2013-08-21 11:49:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1612 (Label size computed wrong within a Form).

2013-07-11 18:14:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1618 (Eliminates 29 warnings of some type in modules from "demos" catalog).

2013-07-11 15:20:54 +0300 Mykola Vshyvkov 

	* Fixed bug #1616 (Eliminates 23 warnings "suggest parentheses around assignment used" in modules from "demos" catalog).

2013-07-11 15:01:50 +0300 Mykola Vshyvkov 

	* Fixed bug #1617 (Eliminates 23 warnings "cast to pointer from integer of different size" in modules from "demos" catalog).

2013-07-11 12:35:30 +0300 Mykola Vshyvkov 

	* Fixed bug #1615 (Eliminates 35 warnings "enumeration value XXX not handled" in modules from "demos" catalog).

2013-07-10 06:55:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1614 (Eliminates 36 warnings "variable may be used uninitialized" in modules from "demos" catalog).

2013-07-10 05:05:28 +0300 Mykola Vshyvkov 

	* Fixed bug #1613 (Eliminates 60 warnings "implicit declaration of function" in modules from "demos" catalog).

2013-07-04 14:08:09 +0300 Mykola Vshyvkov 

	* Fixed bug #1611 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls in modules of "clients" catalog).

2013-07-03 21:27:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1610 (Eliminates 15 warnings of some types in modules of "clients" catalog).

2013-07-03 18:52:21 +0300 Mykola Vshyvkov 

	* Fixed bug #1609 (Eliminates 18 warnings  of "incompatible implicit declaration" in modules of "clients" catalog).

2013-07-03 14:58:22 +0300 Mykola Vshyvkov 

	* Fixed bug #1608 (Eliminates 30 warnings  of "array subscript has type 'char'" in modules of "clients" catalog).

2013-07-03 14:35:04 +0300 Mykola Vshyvkov 

	* Fixed bug #1607 (Eliminates 92 warnings  of "variable may be used uninitialized" in modules of "clients" catalog).

2013-07-03 12:48:55 +0300 Mykola Vshyvkov 

	* Fixed bug #1606 (Eliminates 130 warnings  of "expects argument of type" in modules of "clients" catalog).

2013-07-02 13:40:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 20:52:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 19:07:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1604 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls).

2013-07-01 19:02:25 +0300 Mykola Vshyvkov 

	* Fixed bug #1605 (Eliminates 14 warnings of some types in WML tool).

2013-06-26 23:45:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1603 (Eliminates 18 warnings of some types in some modules).

2013-06-26 19:21:20 +0300 Mykola Vshyvkov 

	* Fixed bug #1602 (Eliminates 6 warnings "implicit declaration of function" in some modules).

2013-06-26 18:36:47 +0300 Mykola Vshyvkov 

	* Fixed bug #1601 (Eliminates 6 warnings: "suggest explicit braces to avoid ambiguous" in some modules).

2013-06-26 17:40:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1600 (Eliminates 26 warnings about using uninitialized variables in some modules).

2013-06-26 15:34:53 +0300 Mykola Vshyvkov 

	* Fixed bug #1599 (Eliminates 14 warnings of incompatible type argument from some modules).

2013-06-25 17:37:02 +0300 Mykola Vshyvkov 

	* Updated ChangeLog.

2013-06-25 13:57:31 +0300 Mykola Vshyvkov 

	* Merge branch 'openmotif_2_3' of ssh://git.code.sf.net/p/motif/code into openmotif_2_3


demos/programs/drag_and_drop/simpledrop.c
/* $XConsortium: simpledrop.c /main/6 1996/04/21 18:50:33 drk $ */
/*
 * Motif
 *
 * Copyright (c) 1987-2012, The Open Group. All rights reserved.
 *
 * These libraries and programs are free software; you can
 * redistribute them and/or modify them under the terms of the GNU
 * Lesser General Public License as published by the Free Software
 * Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * These libraries and programs are distributed in the hope that
 * they will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with these librararies and programs; if not, write
 * to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301 USA
 *
 */
/*
 * HISTORY
 */

/*
 * A Simple Drag And Drop Demo
 *
 *    Created: Vania JOLOBOFF / January 1994
 * ======================================================================
 *
 */

/*
 * ===================== include ========================================
 */

#include <stdlib.h>
#include <stdio.h>
#include <Xm/XmAll.h>

/* ======================================================================
 *  General Types
 */

typedef enum {closed, open} State ;

/*
 * The demo basic data structure
 */

typedef struct _Airport {
   XtAppContext context;
   Display * display;
   int screen_width;
   int screen_height;
   State state;
   Widget main; /* the main window */
   Widget drop_area; /* the drop_area drawing area */
   Widget help_dialog;
   Widget error_dialog;
   Widget warning_dialog;
   Widget msg_area;
   Position drop_x;
   Position drop_y;
} DropDemoRec, *DropDemo;

/* ======================================================================
 * Global constants and variables
 */

Atom StringAtom;
Atom PixmapAtom;
Atom CTAtom;

/* ============================ prototypes  ======================= */

static void Warning(DropDemo this, String s1, String s2);
/* issue a warning to user */

static Widget MenuBarEntry(Widget menubar, String entry, String labels[],
		XtCallbackProc procs[], XtPointer private[], int count);

static void HelpCallback(Widget button, XtPointer client_data, XtPointer cbs);

static void ExitCallback(Widget	button, XtPointer client_data, XtPointer cbs);

static void SomethingDropped(Widget dc, XtPointer client_data,
			     XmDropProcCallbackStruct *call_data);
/* drop proc callback on airport. Something is landing */

static void TransferDone(Widget transfer, XtPointer client_data,
		   Atom * selection, Atom * type, XtPointer value,
		   unsigned long *length, int *format);
/* Xt Selection Callback. Transfer flight number from source */

/* =====================================================================
 * Show application modal warning dialogue box with specified messages
 */

static void Warning(DropDemo this, String s1, String s2)
{
   XmString msg;
   char tmp[512];

   if (s1 == NULL)
     msg = XmStringCreateLocalized(s2);
   else if (s2 == NULL)
     msg = XmStringCreateLocalized(s1);
   else {
      strncat(strcpy(tmp, s1), s2, sizeof(tmp) - strlen(s1));
      msg = XmStringCreateLocalized(tmp);
   }

   if (this->warning_dialog == NULL) {
      Arg args[8];
      int n = 0;

      XtSetArg(args[n], XmNdialogStyle,
	       XmDIALOG_FULL_APPLICATION_MODAL);
      n++;
      this->warning_dialog = XmCreateWarningDialog(this->main,
						"warning", args, n);
   }
   printf("warning %s\n", tmp);
   XtVaSetValues(this->warning_dialog, XmNmessageString, msg, NULL);
/*   XmStringFree(msg); */
   printf("managing dialog\n");
   XtManageChild(this->warning_dialog);
}

/* ===============================================================
 *   Create a cascade and a pulldown entry into a menubar,
 * 	from a list of push button specs.
 *   If the label is NULL, create a separator instead of a push button.
 */

static Widget MenuBarEntry(Widget menubar, String entry, String labels[],
			   XtCallbackProc procs[], XtPointer private[],
			   int count)
{
   Widget menu;
   Widget kid;
   Arg args[8];
   int n;
   int i;
   char name[128];

   n = 0;
   menu = XmCreatePulldownMenu(menubar,
			       strcat(strcpy(name, entry), "Menu"),
			       args, n);
   for( i = 0; i < count; i++) {
      n = 0;
      if (labels[i] == NULL) {
	 kid = XmCreateSeparator(menu, "sep", args, n);
      }
      else {
	 kid = XmCreatePushButton(menu, labels[i], args, n);
	 if (procs[i] != NULL)
	   XtAddCallback(kid, XmNactivateCallback, procs[i], private[i]);
      }
      XtManageChild(kid);
   }
   n = 0;
   XtSetArg(args[n], XmNlabelString, XmStringCreateLocalized(entry)); n++;
   XtSetArg(args[n], XmNsubMenuId, menu); n++;
   kid = XmCreateCascadeButton(menubar, entry, args, n);
   XtManageChild(kid);
   return kid;
}

/* ==============================================================
 *    Main loop
 */

int main(int argc, char *argv[])
{
#define MAIN_CHILDREN		3
#define FORM_CHILDREN		2
#define SetMenuEntry(k,l,p,d) labels[k]=l;procs[k]=p;private[k]=d;
#define NUM_TARGETS 3
   Atom targets[NUM_TARGETS];
   Widget theWidgetRoot;
   Widget main_kid[MAIN_CHILDREN];
   Widget form_kid[FORM_CHILDREN];
   Widget framed[2];
   Widget entry;
   static char myClass[] = "DropDemo";
   Arg args[10];
   int n ;
   String labels[4];
   XtCallbackProc procs[4];
   XtPointer private[4];
   DropDemo this;


/*
 * Initialize
 */
   this = (DropDemo) XtCalloc(sizeof(DropDemoRec), 1);
   this->state = closed;

   theWidgetRoot = XtVaAppInitialize(&this->context, myClass,
				     NULL, 0, &argc, argv, NULL, NULL,
				     XmNallowShellResize, True,
				     NULL);
   this->display = XtDisplay(theWidgetRoot);
   StringAtom = XmInternAtom(this->display, "STRING", False);
   PixmapAtom = XmInternAtom(this->display, "PIXMAP", False);
   CTAtom = XmInternAtom(this->display, "COMPOUND_TEXT", False);
/*
 * Create the widget structure.
 * Create the Main Window, the menubar and the pulldown menus
 * Then a scale and a drawing area inside a frame inside a form
 */
   n = 0;
   this->main = XmCreateMainWindow(theWidgetRoot, "main", args, n);

/*
 * Create the menu bar with a File and Help entries
 */
   n = 0;
   XtSetArg (args[n], XmNscrolledWindowChildType,  XmMENU_BAR); n++ ;
   main_kid[0] = XmCreateMenuBar(this->main, "menubar", args, n);

   n = 0;
   SetMenuEntry(n, "Exit", (XtCallbackProc)ExitCallback, (XtPointer)this );
   n++;
   entry = MenuBarEntry(main_kid[0], "File", labels, procs, private, n);
   n = 0;
   SetMenuEntry(0, "Tutorial", (XtCallbackProc)HelpCallback, (XtPointer)this);
   n++;
   entry = MenuBarEntry(main_kid[0], "Help", labels, procs, private, n);
   n = 0;
   XtSetArg(args[n], XmNmenuHelpWidget, entry); n++;
   XtSetValues(main_kid[0], args, n);
/*
 * Create the work area as a Form containing a Scale and a Frame
 */
   n = 0;
   this->screen_width = WidthOfScreen(XtScreenOfObject(this->main));
   this->screen_height = HeightOfScreen(XtScreenOfObject(this->main));
   XtSetArg(args[n], XmNwidth, this->screen_width*3/7); n++;
   XtSetArg(args[n], XmNheight, this->screen_height*3/7); n++;
   XtSetArg(args[n], XmNresizePolicy, XmRESIZE_ANY); n++;
   XtSetArg (args[n], XmNscrolledWindowChildType, XmWORK_AREA); n++ ;
   main_kid[1] = XmCreateForm(this->main, "work_area", args, n);

   n = 0;
   XtSetArg(args[n], XmNtopAttachment, XmATTACH_FORM); n++;
   XtSetArg(args[n], XmNleftAttachment, XmATTACH_FORM) ; n++;
   XtSetArg(args[n], XmNrightAttachment, XmATTACH_FORM) ; n++;
   XtSetArg(args[n], XmNorientation, XmHORIZONTAL); n++;
   XtSetArg(args[n], XmNvalue, 0); n++;
   XtSetArg(args[n], XmNmaximum, this->screen_width*3/7); n++;
   XtSetArg(args[n], XmNminimum, 0); n++;
   XtSetArg(args[n], XmNshowValue, True); n++;
   XtSetArg(args[n], XmNtitleString,
		      XmStringCreateLocalized("Number To Drag")); n++;
   form_kid[0] = XmCreateScale(main_kid[1], "scale", args, n);

   n = 0;
   XtSetArg(args[n], XmNtopAttachment, XmATTACH_WIDGET) ; n++;
   XtSetArg(args[n], XmNtopWidget, form_kid[0]) ; n++;
   XtSetArg(args[n], XmNbottomAttachment,XmATTACH_FORM); n++;
   XtSetArg(args[n], XmNleftAttachment, XmATTACH_FORM) ; n++;
   XtSetArg(args[n], XmNrightAttachment, XmATTACH_FORM) ; n++;
   form_kid[1] = XmCreateFrame(main_kid[1], "frame", args, n);

   n = 0;
   XtSetArg(args[n], XmNchildType, XmFRAME_TITLE_CHILD); n++;
   framed[0] = XmCreateLabel(form_kid[1], "drop_here", args, n);
/*
 * Create the dropping area
 */
   n = 0;
   this->drop_area =
     framed[1] = XmCreateBulletinBoard(form_kid[1], "da", args, n);
   XtVaSetValues(this->drop_area, XmNuserData, this, NULL);
   n = 0;
   targets[0] = StringAtom;
   targets[1] = PixmapAtom;
   targets[2] = CTAtom;
   XtSetArg(args[n], XmNimportTargets, targets); n++;
   XtSetArg(args[n], XmNnumImportTargets, NUM_TARGETS); n++;
   XtSetArg(args[n], XmNdragProc, NULL); n++;
   XtSetArg(args[n], XmNdropProc, SomethingDropped); n++;
   XmDropSiteRegister(this->drop_area, args, n);

   XtManageChildren(framed, 2);
   XtManageChildren(form_kid, FORM_CHILDREN);

   XtManageChildren(main_kid, 2);
   XtManageChild(this->main);

   XtRealizeWidget(theWidgetRoot);
   XtAppMainLoop(this->context);

   return 0;    /* make compiler happy */
}

/* ===============================================================
 *   The Exit Callback.  Exit properly.
 */

static void ExitCallback(Widget	button, XtPointer client_data, XtPointer cbs)
{
   DropDemo this = (DropDemo) client_data;

   XtCloseDisplay(this->display);
   exit(0);
}

/* ======================================================================
 *  Help Callback. Display help message.
 */

static void HelpCallback(Widget button, XtPointer client_data, XtPointer cbs)
{
   DropDemo this = (DropDemo) client_data;
   static char helpmsg[] =
     "Please drag the value of the scale into the drop area";

  if ( this->help_dialog == NULL ) {
     Arg args[8];
     int n = 0;

     XtSetArg(args[n], XmNmessageString,
	      XmStringCreateLocalized(helpmsg)); n++;
     this->help_dialog = XmCreateInformationDialog(button,
					    "helpdialog", args, n);
  }
  XtManageChild(this->help_dialog);
}

/* ===============================================================
 * Drop proc callback of the drop site.
 * If the drop site is invalid, issue a warning
 * If the drop site is valid, we know the target is at least one of
 * 	String, CT or Pixmap.
 * Pick one of them only in this order of preference.
 */

static void SomethingDropped(Widget dropSite, XtPointer client_data,
			     XmDropProcCallbackStruct *call_data)
{
#define NUMREC 2
   DropDemo this;
   Arg args[8];
   Arg get_args[4];
   Atom * exports;
   XmDropTransferEntryRec target_data[NUMREC];
   unsigned char status = '\0';
   Widget dc = call_data->dragContext;
   int n = 0;
   int i = 0;
   Cardinal num_targets = 0;
   Cardinal num_transfer = 0;

   XtVaGetValues(dropSite, XmNuserData, &this, NULL);
   XtSetArg(get_args[i], XmNexportTargets, &exports);  i++;
   XtSetArg(get_args[i], XmNnumExportTargets, &num_targets); i++;
   XtGetValues(dc, get_args, i);
   if (call_data->dropSiteStatus == XmVALID_DROP_SITE) {
      call_data->operation = XmDROP_COPY;
      this->drop_x = call_data->x;
      this->drop_y = call_data->y;
      for (i = 0; i < num_targets; i++ )
	if (exports[i] == StringAtom
	    || exports[i] == CTAtom
	    || exports[i] == PixmapAtom) {
	   status = XmTRANSFER_SUCCESS;
	   num_transfer = 1;
	   target_data[0].client_data = (XtPointer) this;
	   target_data[0].target = exports[i];
	   XtSetArg(args[n], XmNtransferProc, TransferDone);  n++;
	   XtSetArg(args[n], XmNdropTransfers, target_data);  n++;
	   break;
	}
   }
   else     {
      Warning(this, "Non Identified Object is Dropped. ",
	      num_targets > 0 ? XmGetAtomName(this->display,exports[0]) : "");
      status = XmTRANSFER_FAILURE;
   }
   XtSetArg(args[n], XmNnumDropTransfers, num_transfer);  n++;
   XtSetArg(args[n], XmNtransferStatus, status);  n++;
   XmDropTransferStart(call_data->dragContext, args, n);
#undef NUMREC
}

/* ===============================================================
 * The selection callback proc after drop transfer.
 * Bytes have been received. Check they are OK
 * If data received is corrupted, issue a warning.
 */

static void TransferDone(Widget transfer, XtPointer client_data,
			 Atom * selection, Atom * type, XtPointer value,
			 unsigned long *length, int *format)
{
   Arg args[16];
   int n = 0;
   Widget label;
   DropDemo this = (DropDemo) client_data;
   XmString string = NULL;
   char msg[256];


   if (*format == 8)
     {
	if (*type == StringAtom)
	  string = XmStringCreateLocalized((char *)value);
	else if (*type == CTAtom)
	  string = XmCvtCTToXmString( (char *) value );
      XtSetArg(args[n], XmNlabelString, string); n++;
     }
   else if (*type == PixmapAtom && *format == 32) {
      Pixmap pixmap = (Pixmap) value;
      XtSetArg(args[n], XmNbackgroundPixmap, pixmap); n++;
   }
   else {
      XtVaSetValues(transfer,
		    XmNtransferStatus, XmTRANSFER_FAILURE,
		    XmNnumDropTransfers, 0,
		    NULL);
      sprintf(msg, "Type %s format %d",
	      XmGetAtomName(this->display, *type), *format);
      Warning(this, "Dropped data is corrupted. ", msg);
      return;
   }
   XtSetArg(args[n], XmNx, this->drop_x); n++;
   XtSetArg(args[n], XmNy, this->drop_y); n++;
   label = XmCreateLabel(this->drop_area, "label", args, n);
   XtManageChild(label);
   if (string != NULL)
      XmStringFree(string);
}
